File: src/DFRobot_EOxygenSensor.cpp
/*!
 * @file DFRobot_EOxygenSensor.cpp
 * @brief Define the basic structure of class DFRobot_EOxygenSensor
 * @copyright   Copyright (c) 2010 DFRobot Co.Ltd (http://www.dfrobot.com)
 * @license     The MIT License (MIT)
 * @author      ZhixinLiu(zhixin.liu@dfrobot.com)
 * @version     V1.0
 * @date        2021-12-28
 * @url https://github.com/DFRobot/DFRobot_EOxygenSensor
 */
#include "DFRobot_EOxygenSensor.h"
/*
DFRobot_EOxygenSensor::DFRobot_EOxygenSensor()
{
}

DFRobot_EOxygenSensor::~DFRobot_EOxygenSensor()
{
}
*/


void DFRobot_EOxygenSensor::readKB(void)
{
  uint8_t buf[10] = {0x00};
  readData(25, buf, 6);
  // Serial.print("k  "); Serial.println((float)buf[0] + (float)buf[1]/100.0);
  // Serial.print("b  "); Serial.println((float)buf[2] + (float)buf[3]/100.0);
  // Serial.print("k0  "); Serial.println((float)buf[4] + (float)buf[5]/100.0);
}

float DFRobot_EOxygenSensor::readOxygenConcentration(void)
{
  static float oldVol = 0.0;
  float vol;
  uint8_t buf[10] = {0x00};
  readData(OXYGEN_DATA, buf, 3);
  vol = (float)buf[0] + ((float)buf[1])/10.0 + ((float)buf[2])/100.0;
  if(vol >= 0.00001 && vol <= -0.00001){
    return oldVol;
  }else{
    oldVol = vol;
    return vol;
  }
}

uint8_t DFRobot_EOxygenSensor::readCalibrationState(void)
{
  uint8_t buf[10] = {0x00};
  readData(CALIBRATION_STATE, buf, 1);
  return buf[0];
}

bool DFRobot_EOxygenSensor::calibration_20_9(void)
{
  uint8_t state = 0, data = CALIBRATION_20_9;
  writeData(CALIBRATION_SENSOR, &data, 1);
  delay(2000);
  state = readCalibrationState();
  if(state & CALIBRATION_20_9)  return true;
  return false;
}

bool DFRobot_EOxygenSensor::calibration_99_5(void)
{
  uint8_t state = 0, data = CALIBRATION_99_5;
  writeData(CALIBRATION_SENSOR, &data, 1);
  delay(2000);
  state = readCalibrationState();
  if(state & CALIBRATION_99_5)  return true;
  return false;
}

bool DFRobot_EOxygenSensor::clearCalibration(void)
{
  uint8_t state = 0, data = CALIBRATION_CLEAR;
  writeData(CALIBRATION_SENSOR, &data, 1);
  delay(2000);
  state = readCalibrationState();
  if(state == 0)  return true;
  return false;
}

DFRobot_EOxygenSensor_I2C::DFRobot_EOxygenSensor_I2C(TwoWire *pWire, uint8_t addr)
{
  _pWire = pWire;
  this->_I2C_addr = addr;
}

bool DFRobot_EOxygenSensor_I2C::begin(void)
{
  _pWire->begin();
  _pWire->setClock(100000);
  _pWire->beginTransmission(_I2C_addr);
  
  if(_pWire->endTransmission() == 0){
    return true;
  }else{
    DBG("I2C connect error");
    return false;
  }
}

void DFRobot_EOxygenSensor_I2C::writeData(uint8_t reg, uint8_t *data, uint8_t len)
{
  _pWire->beginTransmission(this->_I2C_addr);
  _pWire->write(reg);
  for(uint8_t i = 0; i < len; i++)
    _pWire->write(data[i]);
  _pWire->endTransmission();
}

uint8_t DFRobot_EOxygenSensor_I2C::readData(uint8_t reg, uint8_t *data, uint8_t len)
{
  int i=0;
  _pWire->beginTransmission(this->_I2C_addr);
  _pWire->write(reg);
  _pWire->endTransmission();
  _pWire->requestFrom((uint8_t)this->_I2C_addr,(uint8_t)len);
  while (_pWire->available()){
    data[i++]=_pWire->read();
  }
  return 0;
}

File: python/raspberrypi/DFRobot_EOxygenSensor.py
# -*- coding:utf-8 -*-
'''!
  @file DFRobot_EOxygenSensor.py
  @note DFRobot_EOxygenSensor Class infrastructure, implementation of underlying methods
  @copyright   Copyright (c) 2010 DFRobot Co.Ltd (http://www.dfrobot.com)
  @license     The MIT License (MIT)
  @author      [ZhixinLiu](zhixin.liu@dfrobot.com)
  @version     V1.0
  @date       2021-12-29
  @url https://github.com/DFRobot/DFRobot_EOxygenSensor
'''
import time
import smbus
import os

## Oxygen data register
OXYGEN_DATA               = 0x10
## calbration state
CALIBRATION_STATE         = 0x13
## calbration sensor
CALIBRATION_SENSOR        = 0x18
## I2C address 
E_OXYGEN_ADDRESS_0        = 0x70
E_OXYGEN_ADDRESS_1        = 0x71
E_OXYGEN_ADDRESS_2        = 0x72
E_OXYGEN_ADDRESS_3        = 0x73

CALIBRATION_20_9          = 0x01
CALIBRATION_99_5          = 0x02
CALIBRATION_CLEAR         = 0x03

class DFRobot_EOxygenSensor(object):
  '''!
    @brief Electrochemical oxygen class
    @details Contain some functions of obtaining/calibrating oxygen data
  '''
  ## iic send buffer
  __txbuf      = [0]    
  def __init__(self, bus):
    self.i2cbus = smbus.SMBus(bus)

  def read_oxygen_concentration(self):
    '''!
      @brief Get oxygen concentration in the air unit VOL
      @return concentration
    '''
    rslt = self.read_reg(OXYGEN_DATA, 3)
    vol = float(rslt[0]) + float(rslt[1]) / 10.0 + float(rslt[2]) / 100.0
    return vol


  def read_calibration_state(self):
    '''!
      @brief Get calibration status
      @return state
    '''
    rslt = self.read_reg(CALIBRATION_STATE, 3)
    return rslt[0]

  def calibration_20_9(self):
    '''!
      @brief Calibrate sensor in air with an O2 concentration of 20.9% Vol
      @return Calibration status
    '''
    self.__txbuf[0] = 1
    self.write_reg(CALIBRATION_SENSOR, self.__txbuf)
    time.sleep(2)
    state = self.read_calibration_state()
    if state & CALIBRATION_20_9 == CALIBRATION_20_9:
      return 1
    else:
      return 0

  def calibration_99_5(self):
    '''!
      @brief Calibrate sensor in 99.5% pure oxygen
      @return Calibration status
    '''
    self.__txbuf[0] = 2
    self.write_reg(CALIBRATION_SENSOR, self.__txbuf)
    time.sleep(2)
    state = self.read_calibration_state()
    if state & CALIBRATION_99_5 == CALIBRATION_99_5:
      return 1
    else:
      return 0

  def clear_calibration(self):
    '''!
      @brief Clear calibration status
      @return Calibration status
    '''
    self.__txbuf[0] = 3
    self.write_reg(CALIBRATION_SENSOR, self.__txbuf)
    time.sleep(2)
    state = self.read_calibration_state()
    if state == 0:
      return 1
    else:
      return 0


class DFRobot_EOxygenSensor_I2C(DFRobot_EOxygenSensor): 
  '''!
    @brief An example of an IIC interface module
  '''

  def __init__(self, bus, addr):
    self.__addr = addr
    super(DFRobot_EOxygenSensor_I2C, self).__init__(bus)

  def write_reg(self, reg, data):
    '''!
      @brief writes data to a register
      @param reg register address
      @param data written data
    '''
    self.i2cbus.write_i2c_block_data(self.__addr, reg, data)

  def read_reg(self, reg, len):
    '''!
      @brief read the data from the register
      @param reg register address
      @param len read datalen
    '''
    while 1:
      try:
        rslt = self.i2cbus.read_i2c_block_data(self.__addr, reg, len)
        return rslt
      except:
        os.system('i2cdetect -y 1')
